File: src/main/java/com/tjh/riskfactor/repo/TaskRepository.java
package com.tjh.riskfactor.repo;

import org.springframework.stereotype.Repository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.jpa.repository.JpaRepository;

import com.fasterxml.jackson.annotation.JsonFormat;

import com.tjh.riskfactor.entity.form.Task;

import java.util.Date;
import java.util.List;

@Repository
public interface TaskRepository extends JpaRepository<Task, Integer> {

    interface TaskBrief {
        Integer getId();
        String getName();
        @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
        Date getMtime();
        String getCenter();
    }

    interface SectionBrief {
        Integer getId();
        String getTitle();
    }

    @Query("select t.id as id, t.name as name, t.mtime as mtime, g.name as center from Task t join t.group g")
    List<TaskBrief> findAllTasks();

    @Query("select s.id as id, s.title as title from Task t join t.sections s where t.id = :id")
    List<SectionBrief> findSectionNamesById(Integer id);

}

File: backend/src/main/kotlin/com/tjh/riskfactor/repo/TaskRepository.kt
package com.tjh.riskfactor.repo

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository

import com.tjh.riskfactor.entity.form.Task
import com.tjh.riskfactor.entity.view.*

private const val TASK_BRIEF_ALL = "select t.id as id, t.name as name, t.mtime as mtime, g.displayName as center from Task t join t.group g "

@Repository
interface TaskRepository : JpaRepository<Task, Int> {

    @Query("select t.mtime as mtime from Task t where t.id = :id")
    fun mtime(id: Int): MtimeView?

    @Query(TASK_BRIEF_ALL + "on t.id = :id")
    fun taskView(id: Int): TaskView?

    @Query(TASK_BRIEF_ALL)
    fun taskViews(): List<TaskView>

}
